src/io/darwin.zig

```diff
@@ -638,19 +607,160 @@ pub const IO = struct {
                 .offset = offset,
             },
             struct {
-                fn doOperation(op: anytype) WriteError!usize {
+                fn do_operation(op: anytype) WriteError!usize {
                     return os.pwrite(op.fd, op.buf[0..op.len], op.offset);
                 }
             },
         );
     }
 
-    pub fn openSocket(family: u32, sock_type: u32, protocol: u32) !os.socket_t {
+    pub const INVALID_SOCKET = -1;
+
+    pub fn open_socket(self: *IO, family: u32, sock_type: u32, protocol: u32) !os.socket_t {
         const fd = try os.socket(family, sock_type | os.SOCK_NONBLOCK, protocol);
-        errdefer os.close(fd);
+        errdefer os.closeSocket(fd);
 
         // darwin doesn't support os.MSG_NOSIGNAL, but instead a socket option to avoid SIGPIPE.
         try os.setsockopt(fd, os.SOL_SOCKET, os.SO_NOSIGPIPE, &mem.toBytes(@as(c_int, 1)));
         return fd;
     }
+
+    pub fn open_dir(dir_path: [:0]const u8) !os.fd_t {
+        return os.openZ(dir_path, os.O_CLOEXEC | os.O_RDONLY, 0);
+    }
+
+    pub fn open_file(
+        self: *IO,
+        dir_fd: os.fd_t,
+        relative_path: [:0]const u8,
+        size: u64,
+        must_create: bool,
+    ) !os.fd_t {
+        // TODO Use O_EXCL when opening as a block device to obtain a mandatory exclusive lock.
```

Comment:
Can we bring back the `assert(relative_path.len > 0);` that we had? Or what is the reason for dropping it?

We should also assert that `size > 0`.

Reply:
Let's also bring back these two assertions, since we always align to sector size regardless of platform:
```zig
        assert(size >= config.sector_size);
        assert(size % config.sector_size == 0);
```

Reply:
They stayed in `Storage.open()` which then calls `IO.open_file()`. We could duplicate them in the IO.open_file() implementations for clarity?

Reply:
I saw the last review comment. Will move the asserts into `IO.open_file`.

Reply:
Great, I was always thinking that we would just use `IO.open_file` directly but didn't make this clear.